ntoskrnl/ex/event.c

```diff
@@ -134,40 +141,41 @@ NtCreateEvent(OUT PHANDLE EventHandle,
                             0,
                             0,
                             (PVOID*)&Event);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("ObCreateObject failed: 0x%X\n", Status);
+        return Status;
+    }
 
-    /* Check for Success */
-    if (NT_SUCCESS(Status))
+    /* Initialize the Event */
+    KeInitializeEvent(Event, EventType, InitialState);
+
+    /* Insert it */
+    Status = ObInsertObject((PVOID)Event,
+                            NULL,
+                            DesiredAccess,
+                            0,
+                            NULL,
+                            &hEvent);
+    if (!NT_SUCCESS(Status))
     {
-        /* Initialize the Event */
-        KeInitializeEvent(Event,
-                          EventType,
-                          InitialState);
-
-        /* Insert it */
-        Status = ObInsertObject((PVOID)Event,
-                                 NULL,
-                                 DesiredAccess,
-                                 0,
-                                 NULL,
-                                 &hEvent);
-
-        /* Check for success */
-        if (NT_SUCCESS(Status))
-        {
-            /* Enter SEH for return */
-            _SEH2_TRY
-            {
-                /* Return the handle to the caller */
-                *EventHandle = hEvent;
-            }
-            _SEH2_EXCEPT(ExSystemExceptionFilter())
-            {
-                /* Get the exception code */
-                Status = _SEH2_GetExceptionCode();
-            }
-            _SEH2_END;
-        }
+        DPRINT1("ObInsertObject failed: 0x%X\n", Status);
+        ObDereferenceObject(Event);
+        return Status;
+    }
+
+    /* Enter SEH for return */
+    _SEH2_TRY
+    {
+        /* Return the handle to the caller */
+        *EventHandle = hEvent;
+    }
+    _SEH2_EXCEPT(ExSystemExceptionFilter())
+    {
+        /* Get the exception code */
+        Status = _SEH2_GetExceptionCode();
```

Comment:
Should we delete the created object if we fail here? (otherwise we "loose" it, the obtained handle (+object reference) being then lost).

`NtCreateTimer` may enjoy a similar treatment as well (including setting the Status on exception):
https://github.com/reactos/reactos/blob/fff6d960003fe79b17d6e52875de60d3964e446b/ntoskrnl/ex/timer.c#L448-L452